### PR TITLE
Cleanup: Remove SetAttribute(string) and AddEvent(name, attributes) overloads on Span

### PR DESCRIPTION
**Pull request by lmolkova**
**SetAttribute** overloads allow to set bool, long, double, int, string and object attributes. 
According to the spec any numeric, boolean or string types are allowed AND arrays of them.

> The attribute value, which is either:
 A primitive type: string, boolean or numeric.
An array of primitive type values. The array MUST be homogeneous, i.e. it MUST NOT contain values of different types.

This makes it impossible to keep up with all possible attribute types through the API. 
So this change **limits SetAttribute to (string, object/long/double/bool) overloads only** and does not allow non-supported attribute types. Boxing is inevitable in any case, since we are storing attributes in IEnumerable<KeyValuePair<string, object>>

Also, we have **3 overloads to add event on span** (from name,name and attributes and from event instance).
This change adds Event ctor with (name, attributes), but limits AddEvent on Span to `AddEvent(name)` and `AddEvent(event)`

### CODE REVIEW

**Comment by SergeyKanzhelev** [comment]
This change prevent us from avoiding box-ing when span is not being collected. I think it may be a problem. What do you think?

**Comment by lmolkova** [comment]
This is the feeback we got from .NET folks:

1. boxing is not a big deal
2. if you're care about perf a lot, you first check for `span.IsRecording` before adding any attributes (for HTTP, it's a few string allocations to fill all the attributes)

**Comment by bruno-garcia** [comment]
I second @MikeGoldsmith on this one

**Comment by lmolkova** [comment]
> I'm not a fan of this change. It's nice the API surface is reduced, but having discrete typed SetAttribute overloads feels nicer and we don't have to type check on every call.

Sure, would extension methods on span solve this concern?

```csharp
public static void SetAttribute(this ISpan span, string key, long value)
{
    if (span.IsRecording) 
        span.SetAttribute(key, value0;
}
```
this will keep abstraction and implementations smaller without any perf implications

**Comment by MikeGoldsmith** [comment]
The extension methods would expose typed overloads but the base implementation would still be exposed presenting `object` as a valid type and it always has do type checking because someone could use the base implementation directly instead of the extension method.

I'm still on the fence with this change -- my preference would be discrete typed methods exposed directly on `Span`.

**Comment by lmolkova** [comment]
@MikeGoldsmith , yes, you are right. 

I'll return back long/double/bool. Are we on the same page (despite spec allowing to have any numeric) that we should not add overloads for byte/etc?

**Comment by MikeGoldsmith** [comment]
Yeah, I think long, double and bool should be enough.

Only reason I can think of for allowing other types (byte, int, etc) would be reduced allocations but we're storing them into `Dictionary<string, object>` anyway, so doubt we'd see much benefit.

File: src/OpenTelemetry.Api/Internal/BlankSpan.cs [hunk 1]
```diff
@@ -66,21 +65,11 @@ public void SetAttribute(string key, double value)
...
-        {
-        }
-
         /// <inheritdoc />
         public void AddEvent(string name)
         {
         }
 
-        /// <inheritdoc />
-        public void AddEvent(string name, IDictionary<string, object> attributes)
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Api/Internal/BlankSpan.cs]: by removing it - an option to `AddEvent` optimization that prevents allocation of an `Event` object disappears. So API forces allocation. Is it intended?

Comment (SergeyKanzhelev) [reply src/OpenTelemetry.Api/Internal/BlankSpan.cs]: Also it makes the constructor that accepts DateTimeOffset more visible. Which can lead to the incorrect use of an API where the timestamp is passed explicitly.

Comment (lmolkova) [reply src/OpenTelemetry.Api/Internal/BlankSpan.cs]: there is a requirement to support lazy events. This also does not prevent allocation (lazy is Func which is usually an allocation). 

I'm still strong that instrumentation that cares about perf MUST use IsRecording flag to optimize. Creating a dictionary of attributes is bad enough before checking for isRecording so it does not make this method optimal anyway.

Preventing allocations by exposing methods on Span mans that Span API explodes to support all possible/popular ways to create events/links/set attributes. I would prefer to stay on the small and clean API (which requires instrumentation to optimize by using  IsRecording)

Comment (lmolkova) [reply src/OpenTelemetry.Api/Internal/BlankSpan.cs]: > Also it makes the constructor that accepts DateTimeOffset more visible. Which can lead to the incorrect use of an API where the timestamp is passed explicitly.

Let's do something about it and make event timestamp usage hard

**Pull request merged**